typescript-rewrite/samlang-core/src/interpreter/expression-interpreter.ts

```diff
@@ -0,0 +1,274 @@
+            return { type: 'bool', value: expression.literal.value };
+        }
+        break;
+      }
+      case 'ThisExpression':
+        return context.localValues.get('this') ?? this.blameTypeChecker('Missing `this`');
+      case 'VariableExpression':
+        return (
+          context.localValues.get(expression.name) ??
+          this.blameTypeChecker(`Missing variable ${expression}`)
+        );
+      case 'ClassMemberExpression':
+        return (
+          context.classes.get(expression.className)?.functions?.get(expression.memberName) ??
+          this.blameTypeChecker()
+        );
+      case 'TupleConstructorExpression':
+        return {
+          type: 'tuple',
+          tupleContent: expression.expressions.map((e) => this.eval(e, context)),
+        };
+      case 'ObjectConstructorExpression': {
+        const objectContent = new Map<string, Value>();
+        expression.fieldDeclarations.forEach((declaration) => {
+          if (declaration.expression) {
+            objectContent.set(declaration.name, this.eval(declaration.expression, context));
+          }
+        });
+        return { type: 'object', objectContent };
+      }
+      case 'VariantConstructorExpression':
+        return { type: 'variant', tag: expression.tag, data: this.eval(expression.data, context) };
+      case 'FieldAccessExpression': {
+        const thisValue = this.eval(expression.expression, context) as ObjectValue;
+        return thisValue.objectContent.get(expression.fieldName) ?? this.blameTypeChecker();
+      }
+      case 'MethodAccessExpression': {
+        const { identifier } = expression.expression.type as IdentifierType;
+        const thisValue = this.eval(expression.expression, context);
+        const methodValue =
+          context.classes.get(identifier)?.methods?.get(expression.methodName) ??
+          this.blameTypeChecker();
+        const newCtx = { ...context };
+        newCtx.localValues = context.localValues.set('this', thisValue);
+        // should context not be readonly?
+        methodValue.context = newCtx;
+        return methodValue;
+      }
+      case 'UnaryExpression': {
+        let v = this.eval(expression.expression);
+        switch (expression.operator) {
+          case '-':
+            v = v as IntValue;
+            return { type: 'int', value: -v.value };
+          case '!':
+            v = v as BoolValue;
+            return { type: 'bool', value: !v.value };
+        }
+        break;
+      }
+      case 'PanicExpression':
+        throw new PanicException((this.eval(expression.expression, context) as StringValue).value);
+      case 'BuiltInFunctionCallExpression': {
+        const argumentValue = this.eval(expression.argumentExpression, context);
+        switch (expression.functionName) {
+          case 'stringToInt': {
+            const value = (argumentValue as StringValue).value;
+            const parsedValue = parseInt(value, 10);
+            if (parsedValue) {
```

Comment:
This is sketchy. If `parsedValue === 0`, it will fail the check. Use `!= null` instead